### PR TITLE
feat(eks): trainium instance types

### LINKED ISSUE #29131: (aws-eks): Neuron device plugin is not installed when instance type is Trainium
### Describe the bug

if instance type is Trainium the neuron device plugin is wrongfully not installed

### Expected Behavior

if instance type is Trainium the neuron device plugin is installed

### Current Behavior

if instance type is Trainium the neuron device plugin is NOT installed

### Reproduction Steps

use an instance of type Trainium

### Possible Solution

_No response_

### Additional Information/Context

Instance types of family Trainium have recently been added here: https://github.com/aws/aws-cdk/blame/main/packages/aws-cdk-lib/aws-ec2/lib/instance-types.ts

BUT:
[packages/aws-cdk-lib/aws-eks/lib/instance-types.ts] does not include them:
export const INSTANCE_TYPES = {
  gpu: ['p2', 'p3', 'g2', 'g3', 'g4'],
  inferentia: ['inf1', 'inf2'],
  graviton: ['a1'],
  graviton2: ['c6g', 'm6g', 'r6g', 't4g'],
  graviton3: ['c7g'],
};

 causing the check in packages/aws-cdk-lib/aws-eks/lib/cluster.ts to fail and the plugin not being installed:

function nodeTypeForInstanceType(instanceType: ec2.InstanceType) {
  return INSTANCE_TYPES.gpu.includes(instanceType.toString().substring(0, 2)) ? NodeType.GPU :
    INSTANCE_TYPES.inferentia.includes(instanceType.toString().substring(0, 4)) ? NodeType.INFERENTIA :
      NodeType.STANDARD;
}

  public addNodegroupCapacity(id: string, options?: NodegroupOptions): Nodegroup {
    const hasInferentiaInstanceType = [
      options?.instanceType,
      ...options?.instanceTypes ?? [],
    ].some(i => i && nodeTypeForInstanceType(i) === NodeType.INFERENTIA);
    if (hasInferentiaInstanceType) {
      this.addNeuronDevicePlugin();
    }
...

### CDK CLI Version

2.128.0

### Framework Version

_No response_

### Node.js Version

v21.6.1

### OS

sonoma 14.3 

### Language

TypeScript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by msambol**
@freschri – It's a little hard to find docs on this but I think this is what you're after?

Closes #29131.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by msambol** [comment]
Exemption request: I don't think integration is required here?

**Comment by freschri** [comment]
> @freschri – It's a little hard to find docs on this but I think this is what you're after?
> 
> Closes #29131.
> 
> _By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

it looks like it thanks. I would just ensure all INFERENTIA used have now a TRINIUM check too

**Comment by msambol** [comment]
@tmokmss Have time for this one? pretty straight-forward..

File: packages/aws-cdk-lib/aws-eks/lib/cluster.ts [hunk 1]
```diff
@@ -2473,7 +2485,8 @@ export enum MachineImageType {
 function nodeTypeForInstanceType(instanceType: ec2.InstanceType) {
   return INSTANCE_TYPES.gpu.includes(instanceType.toString().substring(0, 2)) ? NodeType.GPU :
     INSTANCE_TYPES.inferentia.includes(instanceType.toString().substring(0, 4)) ? NodeType.INFERENTIA :
-      NodeType.STANDARD;
+      INSTANCE_TYPES.trainium.includes(instanceType.toString().substring(0, 4)) ? NodeType.TRAINIUM :
+        NodeType.STANDARD;
```
Comment (tmokmss) [inline packages/aws-cdk-lib/aws-eks/lib/cluster.ts:2489]: This is getting difficult to read. I'd prefer plain if statements.

**Review by tmokmss — APPROVED** [review]
LGTM. since I'm not really familiar with EKS nor Trainium, I commented only about general code quality.

**Comment by msambol** [comment]
> LGTM. since I'm not really familiar with EKS nor Trainium, I commented only about general code quality.

Good feedback, thanks @tmokmss 👍🏼

File: packages/aws-cdk-lib/aws-eks/lib/cluster.ts [hunk 2]
```diff
@@ -1817,11 +1818,13 @@ export class Cluster extends ClusterBase {
    * @param options options for creating a new nodegroup
    */
   public addNodegroupCapacity(id: string, options?: NodegroupOptions): Nodegroup {
-    const hasInferentiaInstanceType = [
+    const hasInferentiaOrTraniumInstanceType = [
```
Comment (shikha372) [inline packages/aws-cdk-lib/aws-eks/lib/cluster.ts:1821]: could we correct this small nit for `hasInferentiaOrTraniumInstanceType` to `hasInferentiaOrTraniumInstanceType`

File: packages/aws-cdk-lib/aws-eks/test/cluster.test.ts [hunk 3]
```diff
@@ -2209,6 +2209,42 @@ describe('cluster', () => {
...
+      });
+      const fileContents = fs.readFileSync(path.join(__dirname, '..', 'lib', 'addons', 'neuron-device-plugin.yaml'), 'utf8');
+      const sanitized = YAML.parse(fileContents);
+
+      // THEN
+      Template.fromStack(stack).hasResourceProperties(eks.KubernetesManifest.RESOURCE_TYPE, {
+        Manifest: JSON.stringify([sanitized]),
+      });
+    });
+    test('trn1n instances are supported', () => {
```
Comment (shikha372) [inline packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:2230]: This unit test seems to be added twice

**Comment by shikha372** [comment]
@msambol PR looks good just some small nits highlighted in above comments, for the integration test could we add simple test like current ones  for inferentia instance type: 
https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/framework-integ/test/aws-eks/test/integ.eks-inference-nodegroup.ts
https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/framework-integ/test/aws-eks/test/integ.eks-inference.ts

File: packages/aws-cdk-lib/aws-eks/test/cluster.test.ts (hunk 3 above)
Comment (msambol) [reply packages/aws-cdk-lib/aws-eks/test/cluster.test.ts:2230]: they look close but one is `trn1` and the other is `trn1n`

**Comment by msambol** [comment]
@paulhcsun possible to get this one in too? I'm going to open a slew of PRs this weekend, want to clean up what's out there now 😄

**Comment by paulhcsun** [comment]
@msambol looks good to me, I'm just waiting on another PR to merge and then i'll update branch and be good to approve

**Review by paulhcsun — APPROVED** [review]

**Pull request merged**